christfellowshipchurch/web-app
# [Events] Chat Direct Messages

# Changes or Updates
- Adds ability to list and view 1:1 DM chats
- Re-organizes some components to break down large `EventChat` file
  - `Tab` components moved into their own file
  - `LiveStreamChat` and `DirectMessagesChat` components created to allow diverging UX needs
  - `DirectMessagesDropdown` created as separate file for clarity
- Adds [`color`](https://www.npmjs.com/package/color) npm package
- Adds values in `theme` to replace `rounded` and `shadow` Bootstrap files
- [Bug fix] Refines Stream chat user management to prevent two calls to `setUser` when logging in from logged out state

### Basecamp Todos
- [**🏕 Display a list of a user's active DMs**](https://3.basecamp.com/3926363/buckets/17566548/todos/3005458757)
- [**🏕 Allow viewing of a specific DM convo**](https://3.basecamp.com/3926363/buckets/17566548/todos/3005465142)

# Tradeoffs/Caveats
- There are rough edges on the design and copy. The `<select>` styling is abysmal, I'm not confident the semi-transparent header is appropriate for CF branding, etc.
- This PR does _not_ add the ability for moderators to initiate DMs —  so DMs must be created manually via code for now.

# Test Instructions
To test this PR, you'll need to workaround the lack of UI ability to create DMs.
The code to do create a 1:1 DM chat is commented out in **src/content-single/EventContentItem/components/EventChat.js** on lines 149–158:
<img src="https://user-images.githubusercontent.com/1812955/93777568-f76d1500-fbf2-11ea-8236-86363b689aa7.png" width="75%"/>
**NOTE** You only need to run this code once to invoke it — I found during testing that subsequent calls will clear the channel history/actually re-create the channel, so if you see messages disappear, that may be why.

### Test logged out/Guest user
1. Log out
1. 🔸 Verify there is no sub-header and that you can only read chats

### Test with 1 active DM
1. Log in as a user that has no DMs yet
1. 🔸 Verify there is no sub-header in chat relating to DMs, and that you can send messages in the LiveStream chat as you
1. Uncomment the code to create a 1:1 DM with someone. Save, make sure the page refreshed, optionally check the console to see a log saying `dmChannelsResponse: ` that should now be an array with 1 item. Re-comment out the code, and save.
1. 🔸 Verify there is now a sub-header in chat saying "DM with (Name) >"
1. Click that link
1. 🔸 Verify you are taken to a sub-screen showing only the 1:1 chat with that user
1. 🔸 Verify chatting with that user works (ideally, log in as the recipient in an Incognito window to see both sides of the chat)
1. Click the "< Back to Chat" link in the sub-header
1. 🔸 Verify you're back in the LiveStream chat, and can seamlessly navigate in/out of the DM

### Test with >= 2 active DMs
1. Create a second DM channel, with a different recipient user (don't forget to re-comment the code out after)
1. 🔸 Verify that you now see an ugly "Direct Messages..." select dropdown menu in the top right of chat, listing both chats
1. 🔸 Verify you can seamlessly navigate between each individual DM chat and back to the LiveStream
1.  🔸Verify the "Back to Chat" text is hidden and that the back button is now just an icon "<"

# Screenshots

|With 1 Active DM|With >= 2 Active DMs|
|---|---|
|![2020-09-19 00 25 13](https://user-images.githubusercontent.com/1812955/93778972-7282fb00-fbf4-11ea-9fed-5497e308e195.gif)|![2020-09-21 09 32 32](https://user-images.githubusercontent.com/1812955/93779012-7dd62680-fbf4-11ea-8733-f15fca117a7d.gif)|

|State|Screenshot|
|---|---|
|**LiveStream chat with 1 active DM**|<img src="https://user-images.githubusercontent.com/1812955/93779106-9f371280-fbf4-11ea-9723-0c4e34c79d37.jpg" width="480" />|
|**LiveStream chat with >= 2 active DMs**|<img src="https://user-images.githubusercontent.com/1812955/93779136-a4945d00-fbf4-11ea-8959-33e5a00a6f61.jpg" width="480" />|
|**DM Chat when only 1 active DM**|<img src="https://user-images.githubusercontent.com/1812955/93779195-b2e27900-fbf4-11ea-8f41-8b2d409d4637.jpg" width="480" />|
|**DM Chat when >= 2 active DMs**|<img src="https://user-images.githubusercontent.com/1812955/93779222-baa21d80-fbf4-11ea-970c-f669a333dcfc.jpg" width="480" />|

# Tests
